src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/decisionmaker/deciders/Publisher.java

```diff
@@ -18,33 +18,65 @@
 import com.amazon.opendistro.elasticsearch.performanceanalyzer.decisionmaker.actions.Action;
 import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.core.NonLeafNode;
 import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.scheduler.FlowUnitOperationArgWrapper;
+
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.Map;
+
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
 public class Publisher extends NonLeafNode<EmptyFlowUnit> {
 
   private static final Logger LOG = LogManager.getLogger(Publisher.class);
+  private final long initTime;
 
   private Collator collator;
   private boolean isMuted = false;
+  private Map<String, Long> actionToExecutionTime;
 
   public Publisher(int evalIntervalSeconds, Collator collator) {
     super(0, evalIntervalSeconds);
     this.collator = collator;
+    this.actionToExecutionTime = new HashMap<>();
+    initTime = Instant.now().toEpochMilli();
+  }
+
+  /**
+   * Returns true if a given {@link Action}'s last execution time was >= {@link Action#coolOffPeriodInMillis()} ago
+   *
+   * <p>If this Publisher has never executed the action, the last execution time is defined as the time that the publisher
+   * object was constructed.
+   *
+   * @param action The {@link Action} to test
+   * @return true if a given {@link Action}'s last execution time was >= {@link Action#coolOffPeriodInMillis()} ago
+   */
+  public boolean isCooledOff(Action action) {
+    long lastExecution = actionToExecutionTime.getOrDefault(action.name(), initTime);
+    long elapsed = Instant.now().toEpochMilli() - lastExecution;
+    if (elapsed >= action.coolOffPeriodInMillis()) {
+      return true;
+    } else {
+      LOG.debug("Action {} still has {} ms left in its cool off period", action.name(),
```

Comment:
Let's keep the logging format consistent within RCA package. Let's add the "Publisher: " at the front of this log

Reply:
I'm against this because it's an anti-pattern in the log4j context. If we want to include class info in logging lines we should update our logging pattern to include %c{1.} for example. See https://logging.apache.org/log4j/2.x/manual/layouts.html

Does anyone else have thoughts on this?

Reply:
I added this in for now